osbuild/util/fscache.py

```diff
@@ -0,0 +1,918 @@
+        entry is completely discarded.
+
+        If the application crashes while holding a staging entry, it will be
+        left behind in the staging directory, but unlocked and marked as stale.
+        Hence, any cache management routine will discard it.
+        """
+
+        assert self._is_active()
+
+        uuidname = None
+        lockfd = None
+
+        try:
+            # Create and lock a new anonymous object in the staging area.
+            (uuidname, lockfd) = self._atomic_dir(self._dirname_stage)
+
+            rpath_data = os.path.join(
+                self._dirname_stage,
+                uuidname,
+                self._dirname_data,
+            )
+
+            # Prepare an empty data directory and yield it to the caller.
+            os.mkdir(os.path.join(self._path_cache, rpath_data))
+            yield rpath_data
+        finally:
+            if lockfd is not None:
+                self._rm_r_object(os.path.join(self._dirname_stage, uuidname))
+                linux.fcntl_flock(lockfd, linux.fcntl.F_UNLCK)
+                os.close(lockfd)
+
+    @contextlib.contextmanager
+    def store(self, name: str):
+        """Store object in cache
+
+        Create a new entry and store it in the cache with the specified name.
+        The entry is first created with an anonymous name and control is yielded
+        to the caller to fill in data. Once control returns, the entry is
+        committed with the specified name.
+
+        The final commit is skipped if an entry with the given name already
+        exists, or its name is claimed for other reasons. Furthermore, the
+        commit is skipped if cache limits are exceeded, or if cache maintenance
+        refuses the commit. Hence, a commit can never be relied upon and the
+        entry might be deleted from the cache as soon as the commit was invoked.
+
+        Parameters:
+        -----------
+        name
+            Name to store the object under.
+        """
+
+        assert self._is_active()
+
+        uuidname = None
+        lockfd = None
+
+        try:
+            # Create and lock a new anonymous object in the staging area.
+            (uuidname, lockfd) = self._atomic_dir(self._dirname_objects)
+
+            rpath_uuid = os.path.join(
+                self._dirname_objects,
+                uuidname,
+            )
+            rpath_data = os.path.join(
+                rpath_uuid,
+                self._dirname_data,
+            )
+            rpath_info = os.path.join(
+                rpath_uuid,
+                self._filename_object_info,
+            )
+            path_uuid = os.path.join(self._path_cache, rpath_uuid)
+            path_data = os.path.join(self._path_cache, rpath_data)
+            path_info = os.path.join(self._path_cache, rpath_info)
+
+            # Prepare an empty data directory and yield it to the caller.
+            os.mkdir(path_data)
+            yield rpath_data
```

Comment:
This returning the name and not the full path is a bit unergonomic, especially since `_path_cache` is private. Maybe we could change this to return the full path _or_ have a read-only `path` property and implement `__fspath__`.

Reply:
I went back and forth on this one. Several comments:

 * Internal helpers like `_atomic_open()` only work on sub-directories of the cache, because they depend on the lockfiles. Hence, I wanted those calls to take relative paths to guard against errors when you accidentally pass paths to other dirs.
 * Whenever you access paths or pass them to external entities, you really want absolute paths to avoid changes in the working-dir to affect the operation. Hence, this is why I added `_path()` as a method to create paths relative to `_path_cache`.
 * When returning paths to the caller, preferably those would be absolute as well. However, this is really awkward if you want to pass those paths back into FsCache, because as noted those calls take relative paths. Hence, I decided to stick to relative paths and rely on post-merge evaluations how to improve this.

I expected this to be awkward to use (and the tests actually end up accessing `_path_cache` all the time...). So I am all open to improve this! Question is, how?

One solution would be to create a helper-class `FsCachePath` or similar, which then implements `__fspath__` and returns the absolute path. But internally, we instead use it to retain the relative path so you can pass it to other FsCache operations. However, it will be awkward to assemble paths with this object, because you *want* to probably assemble relative paths so you can pass it back to `FsCache`. But at the same time you want absolute paths to be able to access them.

So maybe the better approach is to make `__fspath__` return the relative path, but add an `abspath()` function that returns the absolute path? Then all operations would do the right thing, and we could implement `join()` on it which again returns the same object?

Dunno. Maybe the easiest solution is to add `FsCache.abspath()` which takes a relative path and returns an absolute one. Just like `_path()` does, but it always prepends `./objects/<id>/data/`? i will have a closer look at your integration WIP and then see what the callsites actually do.

If you have suggestions, let me know!

Reply:
The two patterns are:
```
         name = self._stack.enter_context(
             self.store.cache.load(self.id)
         )
         self._path = os.path.join(self.store, name)
```
and
```
        with self.store.cache.store(self.id) as name:
             path = os.path.join(self.store, name)
             libc.renameat2(
                 oldpath=self._path.encode(),
                 newpath=path.encode(),
                 flags=libc.RENAME_EXCHANGE,
             )
```

Where `self.store` is `ObjectStore`, which implements `__fspath__` returning `self.store` (unlucky naming), which is exactly the path that was passed to `FsCache`. Ideally `ObjectStore` would not have a copy of it but get it back from `self.cache: FsCache`; also in `Object` would not need to keep a reference to `store` but `cache: FsCache` directly if one could get the path back that was initially passed to it (ideally obviously that would always be an absolute one).
I think it is totally fine for the cache methods to return relative paths as long as one can get also get to the absolute path with just the corresponding `FsCache` around, i.e. `__fspath__` returning an abs path or a `FsCache.abspath(name: str)` would be work and be totally fine I think. Personally I think the `os.PathLike` protocol  , i.e. implementing `__fspath__`, is quite minimalistic and neat but maybe a bit undiscoverable, `FsCache.abspath` is maybe clearer in code. Maybe we could have both.